util/session/sessionmanager.go

```diff
@@ -30,22 +35,118 @@ type SessionManager struct {
 	settingsMgr *settings.SettingsManager
 	client      *http.Client
 	prov        oidcutil.Provider
+	cache       *cacheutil.Cache
+}
+
+// LoginAttempts is a timestamped counter for failed login attempts
+type LoginAttempts struct {
+	// Time of the last failed login
+	LastFailed time.Time `json:"lastFailed"`
+	// Number of consecutive login failures
+	FailCount int `json:"failCount"`
 }
 
 const (
 	// SessionManagerClaimsIssuer fills the "iss" field of the token.
 	SessionManagerClaimsIssuer = "argocd"
 
 	// invalidLoginError, for security purposes, doesn't say whether the username or password was invalid.  This does not mitigate the potential for timing attacks to determine which is which.
-	invalidLoginError  = "Invalid username or password"
-	blankPasswordError = "Blank passwords are not allowed"
-	accountDisabled    = "Account %s is disabled"
+	invalidLoginError    = "Invalid username or password"
+	blankPasswordError   = "Blank passwords are not allowed"
+	accountDisabled      = "Account %s is disabled"
+	usernameTooLongError = "Username is too long (%d bytes max)"
+	// nolint:varcheck,deadcode,unused
+	accountDelayed = "Too many login failures. Login for account %s is delayed by %d seconds"
+)
+
+const (
+	// Maximum length of username, too keep the cache's memory signature low
+	maxUsernameLength = 32
+	// The key prefix to store login attempts in the cache
+	loginAttemptsCacheKey = "session|login.attempts"
+	// The default maximum session cache size
+	defaultMaxCacheSize = 1000
+	// The default number of maximum login failures before delay kicks in
+	defaultMaxLoginFailures = 5
+	// The default delay in seconds to be waited for after login attempts have been reached
+	defaultLoginDelayStart = 3
+	// How many seconds to increase delay after next consecutive login failure
+	defaultLoginDelayIncrease = 2
+	// Default maximum delay in seconds, delay won't be further increased than that
+	defaultLoginDelayMax = 30
+	// The default time in seconds for the failure window
+	defaultFailureWindow = 300
+
+	// environment variables to control rate limiter behaviour
+	envLoginDelayStart    = "ARGOCD_SESSION_FAILURE_DELAY_START"
```

Comment:
We tend to use environment variables for  "hidden" settings which users normally don't need. I feel these settings deserve to be in `argocd-cm` ConfigMap. What do you think?